ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+      ...
+    end
+
+    # Create client with existing connection
+    def initialize(connection, namespace: 'default', ...)
+      ...
+    end
+
+    # Duped hash of keyword args of constructor
+    def options
+      ...
+    end
+
+    # Returns workflow handle
+    def start_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      ...
+    )
+      ...
+    end
+
+    # Returns result
+    def execute_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      ...
+    )
+      ...
+    end
+
+    # Returns workflow handle
+    def workflow_handle(id, run_id: nil, ...)
+      ...
+    end
+
+    # Lots more
+  end
+
+  class Client
+    class Connection
+      # Perform connection
+      def initialize(target_host:, runtime: nil, ...)
+        ...
+      end
+
+      # Duped hash of keyword args of constructor
+      def options
+        ...
+      end
+
+      # Access to raw workflow service
+      def workflow_service
+        ...
+      end
+    end
+
+    class WorkflowHandle
+      # Wait for result
+      def result(follow_runs: true)
+        ...
+      end
+
+      # Wait for result
+      def result(follow_runs: true)
+        ...
+      end
+
+      # Lots more
+    end
+  end
+end
+```
+
+* There are many things missing here, but the idea is clear.
+* We expect most users and most samples to just use `Client.connect`.
+  * 💭 Why not make users create a connection themselves instead of maintaining similar options sets across
+    `Connection.initialize` and `Client.connect`? We have found most people only need a single client per connection.
+* Users that want raw service access need to use `client.connection.workflow_service` similar to .NET.
+* Workflow handles exist with all relevant calls.
+* Note that we prefer large sets of keyword arguments over hashes or options/config objects.
+  * 💭 Why? This is the Ruby preference and clearer to users.
+* Must make clear in docs which things issue an RPC call (e.g. start workflow) vs ones that don't (e.g. get workflow
+  handle).
+
+### Workflows in Client
+
+Early thoughts on workflows are that they'll be Ractors. For the purposes of this phase, all that may matter is how
+workflows are defined so they can be used in a type-safe/custom-named way from a client. All SDKs support defining
+workflows without their body and Ruby should as well.
+
+Most of this TBD until next phase, but for early designs for referencing from clients, this probably looks like:
+
+```ruby
+class MyWorkflow < Temporalio::Workflow
+  workflow_query_attr_reader :some_other_query
+
+  def execute(arg)
+    raise NoMethodError
+  end
+
+  workflow_signal
+  def some_signal(arg)
+    raise NoMethodError
+  end
+
+  workflow_query 'custom-query-name'
+  def some_query(arg)
+    raise NoMethodError
+  end
+end
+```
+
+And client usage may look like:
+
+```ruby
+# Connect client
+client = Client.connect('localhost:7233')
```

Comment:
namespace defaults to `default`?

Reply:
Yes

Reply:
This can definitely be argued, but I think enough of our SDKs default to "default" and some prod clusters do use "default" that yeah it's not a big deal.

Reply:
We are going to update this to require the namespace. The default of `default` is unreasonable for cloud and many others.